common/flusher/flusher.go

```diff
@@ -0,0 +1,274 @@
+	}
+	noopItemWriterImpl[T any] struct {
+		logger log.Logger
+		sync.Mutex
+	}
+)
+
+const (
+	bufferCapacityGapPercent = 0.9
+)
+
+func NewFlusher[T any](
+	bufferCount int,
+	bufferItemCapacity int,
+	flushDuration int,
+	writer ItemWriter[T],
+	logger log.Logger,
+) Flusher[T] {
+	return &flusherImpl[T]{
+		status:                 common.DaemonStatusInitialized,
+		flushDuration:          flushDuration,                                               // time waited after first item insertion before flushing the buffer
+		bufferCount:            bufferCount,                                                 // no of total flush buffers
+		bufferItemCapacity:     bufferItemCapacity,                                          // buffer size, will flush a buffer once no of items added to the buffer nears this limit
+		bufferNearItemCapacity: int(bufferCapacityGapPercent * float64(bufferItemCapacity)), // ^ will flush buffer once buffer size hits this number
+		flushTimer:             nil,
+		flushNotifierChan:      make(chan struct{}, 1),
+		writer:                 writer,
+		flushItemsBuffer:       make([]FlushItem[T], 0, bufferItemCapacity),
+		fullBufferChan:         make(chan []FlushItem[T], bufferCount),
+		freeBufferChan:         make(chan []FlushItem[T], bufferCount),
+		logger:                 logger,
+		shutdownChan:           channel.NewShutdownOnce(),
+	}
+}
+
+func NewNoopItemWriter[T any](logger log.Logger) ItemWriter[T] {
+	return &noopItemWriterImpl[T]{
+		logger: logger,
+	}
+}
+
+func (iwi *noopItemWriterImpl[T]) Write(flushItem ...FlushItem[T]) error {
+	return nil
+}
+
+func (f *flusherImpl[T]) Start() {
+	if !atomic.CompareAndSwapInt32(
+		&f.status,
+		common.DaemonStatusInitialized,
+		common.DaemonStatusStarted,
+	) {
+		return
+	}
+
+	f.flushTimer = time.NewTimer(time.Duration(f.flushDuration))
+	f.flushTimer.Stop() // Stop the timer after creation since we only want timer to start running upon first item insertion
+	f.initFreeBuffers(f.bufferCount, f.bufferItemCapacity)
+
+	go f.timerHandler()
+	go f.fullBufferChanHandler()
+}
+
+func (f *flusherImpl[T]) Stop() {
+	if !atomic.CompareAndSwapInt32(
+		&f.status,
+		common.DaemonStatusStarted,
+		common.DaemonStatusStopped,
+	) {
+		return
+	}
+
+	f.shutdownChan.Shutdown()
+
+	for _, flushItem := range f.flushItemsBuffer {
+		flushItem.fut.Set(struct{}{}, serviceerror.NewCanceled("Unable to write item due to Stop invocation"))
+	}
+}
+
+func (f *flusherImpl[T]) initFreeBuffers(bufferCount int, bufferCapacity int) {
+	for i := 0; i < bufferCount-1; i++ { // -1 since flushItemsBuffer counts as the first free buffer
+		f.freeBufferChan <- make([]FlushItem[T], 0, bufferCapacity)
+	}
+}
+
+func (f *flusherImpl[T]) fullBufferChanHandler() {
+chanLoop:
+	for {
+		select {
+		case fullBuffer := <-f.fullBufferChan:
+			f.flushBuffer(fullBuffer)
+			freeBuffer := fullBuffer[:0]
+			f.freeBufferChan <- freeBuffer
+		case <-f.shutdownChan.Channel():
+			f.Stop()
```

Comment:
Won't this cause a loop? f.shutdownChan.Shutdown() is called in f.Stop(). Besides, f.shutdownChan isn't made available outside this class so the only way this Channel is triggered is from f.Stop(). Therefore, there is no need to call f.Stop() from here. Does that make sense?